### PR TITLE
`First`/`Last` support in Snowflake, and partially in SQLite

### LINKED ISSUE #10411: Support for `First`/`Last` aggregations in Snowflake backend
As part of my work on #9486 I did not manage to easily port our code to support the First and Last aggregations. So I'm creating this ticket to investigate it further.

The first step will be to revert disabling the tests and aggregators:
```diff
Subject: [PATCH] fixing COUNT DISTINCT and trying to fix FIRST
---
Index: distribution/lib/Standard/Snowflake/0.0.0-dev/src/Internal/Snowflake_Dialect.enso
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/distribution/lib/Standard/Snowflake/0.0.0-dev/src/Internal/Snowflake_Dialect.enso b/distribution/lib/Standard/Snowflake/0.0.0-dev/src/Internal/Snowflake_Dialect.enso
--- a/distribution/lib/Standard/Snowflake/0.0.0-dev/src/Internal/Snowflake_Dialect.enso	(revision c03d9fbd541dcee482170df2370cec7031286267)
+++ b/distribution/lib/Standard/Snowflake/0.0.0-dev/src/Internal/Snowflake_Dialect.enso	(date 1719836556251)
@@ -262,7 +262,7 @@
     date_ops = [make_extract_as_int "year", make_extract_as_int "quarter", make_extract_as_int "month", make_extract_as_int "week", make_extract_as_int "day", make_extract_as_int "hour", make_extract_as_int "minute", make_extract_fractional_as_int "second", make_extract_fractional_as_int "millisecond" modulus=1000, make_extract_fractional_as_int "microsecond" modulus=1000, ["date_add", make_date_add], ["date_diff", make_date_diff], ["date_trunc_to_day", make_date_trunc_to_day]]
     special_overrides = []
     other = [["RUNTIME_ERROR", make_runtime_error_op]]
-    my_mappings = text + counts + stats + first_last_aggregators + arith_extensions + bool + date_ops + special_overrides + other
+    my_mappings = text + counts + stats + arith_extensions + bool + date_ops + special_overrides + other
     Base_Generator.base_dialect . extend_with my_mappings

Index: test/Snowflake_Tests/src/Snowflake_Spec.enso
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/Snowflake_Tests/src/Snowflake_Spec.enso b/test/Snowflake_Tests/src/Snowflake_Spec.enso
--- a/test/Snowflake_Tests/src/Snowflake_Spec.enso	(revision c03d9fbd541dcee482170df2370cec7031286267)
+++ b/test/Snowflake_Tests/src/Snowflake_Spec.enso	(date 1719836426136)
@@ -568,7 +568,7 @@
     Common_Spec.add_specs suite_builder prefix create_connection_fn
 
     common_selection = Common_Table_Operations.Main.Test_Selection.Config supports_case_sensitive_columns=True order_by_unicode_normalization_by_default=True allows_mixed_type_comparisons=False fixed_length_text_columns=True removes_trailing_whitespace_casting_from_char_to_varchar=True supports_decimal_type=True supported_replace_params=supported_replace_params run_advanced_edge_case_tests_by_default=False
-    aggregate_selection = Common_Table_Operations.Aggregate_Spec.Test_Selection.Config first_last_row_order=False aggregation_problems=False
+    aggregate_selection = Common_Table_Operations.Aggregate_Spec.Test_Selection.Config first_last=False first_last_row_order=False aggregation_problems=False
     agg_in_memory_table = ((Project_Description.new enso_dev.Table_Tests).data / "data.csv") . read
 
     agg_table_fn = _->
```

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

- Closes #10411 
- Also a similar technique allowed to add partial support for SQLite as well.
	- Only `include_nothing=True` case is still not supported as the technique used for Snowflake did not port well to SQLite and I did not find yet a satisfactory workaround (it is doable, but too ugly to justify it for now IMHO).

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

### CODE REVIEW

**Comment by radeusgd** [comment]
Snowflake test run scheduled: https://github.com/enso-org/enso/actions/runs/10528779943

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Extra Nightly Tests · enso-org/enso@8979be7</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

**Comment by radeusgd** [comment]
The problem with SQLite and `ignore_nothing=True` was that for some reason, SQLite does not support `FILTER` for some window functions:
```
SELECT ROW_NUMBER() FILTER (WHERE a IS NOT NULL) OVER (ORDER BY b)  FROM test;
```
fails with
```
SQLITE_ERROR: FILTER clause may only be used with aggregate window functions
```

Neither does Snowflake have a `MIN_BY`/`MAX_BY` pair of functions like the ones used in Snowflake - if it had these we could do the filtering at the next layer and for `First` find the first non-null value with lowest row number. But don't have such tools, so we need to rely on `row_number = 1` check, but then the row numbers must only be assigned to already non-null values.

There are workaround - for example we can compute row number for all rows in query 1, then compute the MIN row number among non-null rows in query 2 and then use that for filtering just that element in query 3 (so we'd use condition `row_number = min_non_null_number` instead of `row_number = 1`). But that gives us 3 levels of query nesting, instead of the 2 levels that we have already - and I did not like that so I decided to withdraw from going into this rabbit hole, especially as the original ticket was about Snowflake support, and getting it to work in SQLite was supposed to be a nice by-product. If we ever really need to be able to get it to work in SQLite, we can add some special handling, but it may have some trade-offs that I did not want to set right now.

File: distribution/lib/Standard/Snowflake/0.0.0-dev/src/Internal/Snowflake_Dialect.enso [hunk 1]
```diff
@@ -310,6 +309,40 @@ type Snowflake_Dialect
     should_check_table_integrity_at_beginning_of_transaction self -> Boolean =
         False
 
+    ## PRIVATE
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Snowflake/0.0.0-dev/src/Internal/Snowflake_Dialect.enso:312]: Maybe a brief explanation here of why Snowflake requires this custom behavior.

**Review by GregoryTravis — APPROVED** [review]

**Comment by radeusgd** [comment]
In f1faccf2544bee5baf3899755735160ff80b119d I tried using `FIRST_VALUE` as window function in nested query and then extracting it using `ANY_VALUE` as aggregate. However, I ran into some tests failing - so I gave up and reverted it - current implementation works OK, so let's just keep it.

**Comment by radeusgd** [comment]
New run after subsequent changes/merge: https://github.com/enso-org/enso/actions/runs/10578716122

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Extra Nightly Tests · enso-org/enso@84abd84</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

> Edit: the tests have passed :)

**Review by jdunkerley — APPROVED** [review]

**Pull request merged**